solr/core/src/java/org/apache/solr/search/grouping/distributed/shardresultserializer/SearchGroupsResultTransformer.java

```diff
@@ -34,17 +41,37 @@
 /**
  * Implementation for transforming {@link SearchGroup} into a {@link NamedList} structure and visa versa.
  */
-public class SearchGroupsResultTransformer implements ShardResultTransformer<List<Command>, Map<String, SearchGroupsFieldCommandResult>> {
+public abstract class SearchGroupsResultTransformer implements ShardResultTransformer<List<Command>, Map<String, SearchGroupsFieldCommandResult>> {
```

Comment:
An existing public class becoming `abstract` here 'jumped out' at me a little since it could be considered to be breaking backwards compatibility?

The https://github.com/cpoerschke/lucene-solr/commit/7e47e01ba27d3cd39daa288e60c61316b1738d67 commit explores removal/replacement of the `SearchGroupsResultTransformer.getInstance` wrapper and a next step could be for the proposed `SearchGroupsResultTransformer.SkipSecondStepSearchResultResultTransformer` inner class to become a `SkipSecondStepSearchResultResultTransformer.java` class of its own and the proposed `SearchGroupsResultTransformer.DefaultSearchResultResultTransformer` inner class to _not_ be created i.e. `SearchGroupsResultTransformer` here to not become abstract and its constructor to remain public. This of course assumes that `SearchGroupsResultTransformer` would require only simple changes to make it extensible and that then `SkipSecondStepSearchResultResultTransformer` could extend it.

(Note that the https://github.com/cpoerschke/lucene-solr/commit/7e47e01ba27d3cd39daa288e60c61316b1738d67 commit is on https://github.com/cpoerschke/lucene-solr/tree/github-bloomberg-SOLR-11831-cpoerschke-5 branch and that branch includes https://github.com/cpoerschke/lucene-solr/tree/github-bloomberg-SOLR-11831-cpoerschke-4 changes for convenience only i.e. not out of necessity.)

What do you think?

Reply:
It looks a lot better, I merged it with the previous commit. 

One comment: instead of having `SearchGroupsResultTransformer` implementing the default behavior and SkipSecondStepSearchResultResultTransformer, I would compose the behavior: One class for the `default` behaviur`, and one for the `skipsStep` behavior. I'll try to sketch it in a PR to this branch.

Reply:
> ... instead of having `SearchGroupsResultTransformer` implementing the default behavior and SkipSecondStepSearchResultResultTransformer, I would compose the behavior ...

What would happen to `SearchGroupsResultTransformer` itself? Just asking with backwards compatibility in mind.

Reply:
`SearchGroupsResultsTransformer` shouldn't change, it would just delegate the behaviour to a different class according to the `skip` flag.

Reply:
https://github.com/bloomberg/lucene-solr/pull/231 sketches the idea, what do you think?

Reply:
> bloomberg#231 sketches the idea, what do you think?

A pull request for a pull request, or a diff for a diff, at first glance that spooked me a little :-)

And at second glance I skipped 'the middle bit' and looked end-to-end https://github.com/apache/lucene-solr/compare/1071d093360b2c5869a918de743c7089952094f4...5a326a19eb7aa92754d6ccf7b321d3c04d3b9f50 with focus on the `SearchGroupsResultTransformer` class.

There seemed to be a relatively noticable amount of code similarity or duplication between `DefaultSearchResultResultTransformer` and `SkipSecondStepSearchResultResultTransformer` and so i took a step back and considered what the _difference_ (conceptually) between the two transformers is, with two aha! moments:
1. the `getConvertedSortValues` method used by `SkipSecondStepSearchResultResultTransformer` is pretty similar to the `ShardResultTransformerUtils.marshalSortValue` from https://issues.apache.org/jira/browse/SOLR-9890 in Solr 6.5
2. the `SkipSecondStepSearchResultResultTransformer.serializeSearchGroup` method appears to construct exactly what `DefaultSearchResultResultTransformer.serializeSearchGroup` constructs and then it 'wraps' it together with an id and a score in a 'group info' object

The https://github.com/cpoerschke/lucene-solr/commit/10fbfd1dcf16065688c3610b26a55f2aa9c99f8a commit sketches how a `SearchGroupsResultTransformer.serializeOneSearchGroup` method could be factored out. What do you think?

(If factoring out such a method makes sense then I could (next week) attempt a similar approach for the `transformToNative` method. And if that works out good we could revisit how the `[Default|SkipSecondStep]SearchResultResultTransformer` class trio is arranged in terms of hierarchy? And if it does not work out good then that would be insightful too in terms of why etc.)

Reply:
Thanks @cpoerschke I applied you commit (with minor changes) in 9480482166b05d04eb2bb55227baf4f3c6549ab9 and attempted a similar approach for `transformToNative`  in 0eb206bf6f635ba2e7514efa369e09137911b825. Now there is no duplication, but I think we still need to revisit `[Default|SkipSecondStep]SearchResultResultTransformer`... I'm not convinced about SkipSecondStep inheriting and overriding `Default`

Reply:
Thanks @diegoceccarelli for applying the serialise-one-group commit and for attempting a similar approach for the deserialisation! As you say, now there is no duplication though returning to this after a little while it seems to me at least a little bit tricky to understand where the boundaries are between the refactoring and the changes related to the skip-second-step logic i.e. to see that the refactoring does not break anything and to see that and why the skip-second-step logic will work as intended. Against that background https://issues.apache.org/jira/browse/SOLR-13585 proposes to do a pure factoring out of two methods first (albeit with foresight of the envisaged code changes here) and then hopefully the skip-second-step logic addition would be clearer to see. Admittedly though such a split approach would make for a mildly messy rebase for the PR branch here. What do you think?

Reply:
Thanks @cpoerschke  https://issues.apache.org/jira/browse/SOLR-13585 Looks good to me, I would merge it and then rebase this code on top of it..

Reply:
Thanks @diegoceccarelli for the SOLR-13585 review! I'll aim to commit it on Monday then. Rebasing on top of that will also pick up other grouping related changes that happened recently-ish i.e. they will be good to have as well.

Reply:
SOLR-13585 is now committed to master and branch_8x branches; apologies for the delay, the 8.2.0 release process was already underway and I was waiting for the 8.3.0 section to be ready in the solr/CHANGES.txt file.